src/framework/components/element/markup.js

```diff
@@ -0,0 +1,417 @@
+                        break;
+                    default:
+                        this._error = "expected close bracket or identifier";
+                        return false;
+                }
+                token = this._scanner.read();
+            }
+        }
+    });
+
+    // copy the contents of source into target and map the values with the
+    // provided mapping table.
+    function mapAndMerge(target, source) {
+        for (var key in source) {
+            if (!source.hasOwnProperty(key)) {
+                continue;
+            }
+            var value = source[key];
+            if (value instanceof Object) {
+                if (!target.hasOwnProperty(key)) {
+                    target[key] = { };
+                }
+                mapAndMerge(target[key], source[key]);
+            } else {
+                target[key] = value;
+            }
+        }
+    }
+
+    // get the markup tags which apply to the given symbol. all tags overlapping the
+    // symbol must be composed into a single structure.
+    function resolveMarkupTags(tags, symbolIndex) {
+        var result = { };
+        for (var index = 0; index < tags.length; ++index) {
+            var tag = tags[index];
+            if (tag.start <= symbolIndex &&
+                (tag.end === null || tag.end > symbolIndex)) {
+                var tmp = { };
+                tmp[tag.name] = { value: tag.value, attributes: tag.attributes };
+                mapAndMerge(result, tmp);
+            }
+        }
+        return result;
+    }
+
+    // evaluate the list of symbols, extract the markup tags and return an
+    // array of symbol and tag pairs
+    function evaluateMarkup(symbols) {
+        // log scanner output
+        // console.info((new Scanner(symbols)).debugPrint());
+
+        var parser = new Parser(symbols);
+        var stripped_symbols = [];
+        var tags = [];
+
+        if (!parser.parse(stripped_symbols, tags)) {
+            console.warn(parser.error());
+            return {
+                symbols: symbols,
+                tags: null
+            };
+        }
+
+        // if any tags were not correctly closed, return failure
+        var invalidTag = tags.find(function (t) {
+            return t.end === null;
+        });
+
+        if (invalidTag) {
+            console.warn("Markup error: found unclosed tag='" + invalidTag.name + "'");
+            return {
+                symbols: symbols,
+                tags: null
+            };
+        }
+
+        // revolve tags per-character
+        var resolved_tags = null;
+        if (tags.length > 0) {
+            resolved_tags = [];
+            for (var index = 0; index < stripped_symbols.length; ++index) {
+                resolved_tags.push(resolveMarkupTags(tags, index));
+            }
+        }
+
+        return {
+            symbols: stripped_symbols,
+            tags: resolved_tags
+        };
+    }
+
+    return {
+        evaluateMarkup: evaluateMarkup
```

Comment:
One comment I have about this is that it is a little uncommon in the engine to expose functions like that in the pc namespace. It would not stand out as much if this was a method of an exposed class

Reply:
Ok, I will expose a class instead, but not sure if that's what you mean. Please check what I will have done.

Reply:
I guess what I meant was having a class (like Markup although maybe the name Markup is a bit generic?) and having a prototype function for that class and then creating an instance of that class and calling its method. e.g. in the text-element file at the top have an instance of pc.Markup and then call markup.evaluateMarkup on that instance. Static functions are allowed it's just not very common and classes with prototype methods might be a little easier to unit test?